src/Microsoft.AspNet.Mvc.Core/ActionConstraints/ConsumesAttribute.cs

```diff
@@ -0,0 +1,153 @@
+    /// Specifies the allowed content types which can be used to select the action based on request's content-type.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+    public class ConsumesAttribute : Attribute, IResourceFilter, IConsumesActionConstraint
+    {
+        /// <summary>
+        /// Creates a new instance of <see cref="ConsumesAttribute"/>.
+        /// </summary>
+        public ConsumesAttribute([NotNull] string contentType, params string[] otherContentTypes)
+        {
+            ContentTypes = GetContentTypes(contentType, otherContentTypes);
+        }
+
+        // The value used is a non default value so that it avoids getting mixed with other action constraints
+        // with default order.
+        /// <inheritdoc />
+        int IActionConstraint.Order { get; } = 200;
+
+        /// <inheritdoc />
+        public IList<MediaTypeHeaderValue> ContentTypes { get; set; }
+
+        /// <inheritdoc />
+        public void OnResourceExecuting([NotNull] ResourceExecutingContext context)
+        {
+            MediaTypeHeaderValue requestContentType = null;
+            MediaTypeHeaderValue.TryParse(context.HttpContext.Request.ContentType, out requestContentType);
+            if (context.ActionDescriptor.FilterDescriptors.Last(
+                filter => filter.Filter is IConsumesActionConstraint).Filter == this)
+            {
+                // Only execute if this is the last filter before calling the action.
+                // This ensures that we only run the filter which is closest to the action.
+                if (requestContentType != null &&
+                    !ContentTypes.Any(contentType => contentType.IsSubsetOf(requestContentType)))
+                {
+                    context.Result = new UnsupportedMediaTypeResult();
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public void OnResourceExecuted([NotNull] ResourceExecutedContext context)
+        {
+        }
+
+        private bool IsApplicable(ActionConstraintContext context)
+        {
+            // If there are multiple IConsumeActionConstraints which are defined at the class and
+            // at the action level, the one closest to the action overrides the others. To ensure this
+            // we take advantage of the fact that ConsumesAttribute is both an IActionFilter and an
+            // IConsumeActionConstraint. Since filterdescriptor collection is ordered (the last filter is the one
+            // closest to the action), we apply this constraint only if there is no IConsumeActionConstraint after this.
+            return context.CurrentCandidate.Action.FilterDescriptors.Last(
+                filter => filter.Filter is IConsumesActionConstraint).Filter == this;
+        }
+
+        public bool Accept(ActionConstraintContext context)
+        {
+            // First check to see if this is applicable
```

Comment:
the comment is kinda useless. Consider renaming the method, or moving the internal comment here